src/cluster_legacy.c

```diff
@@ -3254,11 +3260,14 @@ int clusterProcessPacket(clusterLink *link) {
                      * to reduce the amount of time needed to stabilize the shard ID. */
                     clusterNode *node = createClusterNode(NULL, CLUSTER_NODE_HANDSHAKE);
                     if (nodeIp2String(node->ip, link, hdr->myip) != C_OK) {
-                        /* We cannot get the IP info from the link, it probably means the connection is closed. */
-                        serverLog(LL_NOTICE, "Closing link even though we received a MEET packet on it, "
-                                             "because the connection has an error");
+                        /* Unable to retrieve the node's IP address from the connection. Without a
+                         * valid IP, the node becomes unusable in the cluster. This failure might be
+                         * due to the connection being closed. To avoid leaving the cluster in an
+                         * inconsistent state, we close the link and free the node. */
+                        serverLog(LL_NOTICE, "Closing link and freeing node due to failure to retrieve IP "
+                                             "from the connection, possibly caused by a closed connection.");
                         freeClusterLink(link);
-                        freeClusterNode(node);
+                        freeClusterNode(node, 0);
                         return 0;
                     }
```

Comment:
This change introduces additional state to `freeClusterNode` in a way that might be hard to follow. I'd suggest optimizing for clarity by opting for a slightly less efficient approach (i.e., an extra memcpy) to improve intuitiveness.

```suggestion
char ip_buf[NET_IP_STR_LEN];
if (nodeIp2String(ip_buf, link, hdr->myip) != C_OK) {
    freeClusterLink(link);
    return 0;
}
clusterNode *node = createClusterNode(NULL, CLUSTER_NODE_HANDSHAKE);
memcpy(node->ip, ip_buf, sizeof(ip_buf));
```

Reply:
right! it is much better